src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -83,8 +83,16 @@ public override void CopyTo(int sourceIndex, char[] destination, int destination
         public override SourceText WithChanges(IEnumerable<TextChange> changes)
         {
             // compute changes against newText to avoid capturing strong references to this ChangedText instance.
-            var changed = (ChangedText)_newText.WithChanges(changes);
-            return new ChangedText(this, changed._newText, changed._changes);
+            // note: _newText is never a ChangeText, so if WithChanges is a no-op the result will not be a ChangedText
+            var changed = _newText.WithChanges(changes) as ChangedText;  
+            if (changed != null)
+            {
+                return new ChangedText(this, changed._newText, changed._changes);
+            }
+            else 
+            {
+                return this;
```

Comment:
The no-op pattern is common throughout the Roslyn API.  Typically API's that return changed instances return the original instance if the change is not really a change.